src/Network/Minio/ListOps.hs

```diff
@@ -25,34 +25,54 @@ import           Lib.Prelude
 import           Network.Minio.Data
 import           Network.Minio.S3API
 
--- | List objects in a bucket matching the given prefix. If recurse is
--- set to True objects matching prefix are recursively listed.
-listObjects :: Bucket -> Maybe Text -> Bool -> C.ConduitM () ObjectInfo Minio ()
+-- | Represents a list output item - either an object or an object
+-- prefix (i.e. a directory).
+data ListItem = ListItemObject ObjectInfo
+              | ListItemPrefix Text
+    deriving (Show, Eq)
+
+-- | @'listObjects' bucket prefix recurse@ lists objects in a bucket
+-- similar to a file system tree traversal.
+--
+-- If @prefix@ is not 'Nothing', only items with the given prefix are
+-- listed, otherwise items under the bucket are returned.
+--
+-- If @recurse@ is set to @True@ all directories under the prefix are
+-- recursively traversed and only objects are returned.
+--
+-- If @recurse@ is set to @False@, objects and directories immediately
+-- under the given prefix are returned (no recursive traversal is
+-- performed).
+listObjects :: Bucket -> Maybe Text -> Bool -> C.ConduitM () ListItem Minio ()
 listObjects bucket prefix recurse = loop Nothing
   where
-    loop :: Maybe Text -> C.ConduitM () ObjectInfo Minio ()
+    loop :: Maybe Text -> C.ConduitM () ListItem Minio ()
```

Comment:
why not have ObjectInfo to have a method to indicate if its a directory not? like isDir() ?

Reply:
That is not a good approach - if a value of ObjectInfo type is actually a directory but is used like an ObjectInfo, we would get unexpected behaviour. To avoid doing this, users would be forced to check isDir every time. Instead making a directory value explicit makes that problem go away at compile time - now all values of ObjectInfo type are actually info about objects.

Reply:
yeah but a prefix is also an object @donatello ListPrefix is not meaningful.  An application would be able to easily use the ObjectInfo as is without worrying about multiple types. 

Explicit type is generally good IMO but in this case, we are just exposing the Listing implementation details which is not needed. The listObjects should only return ObjectInfo() where the objectInfo describes itself like FileInfo when you navigate a directory.

Reply:
Other advantages are writing a File Tree Walk like implementations using ListObjects() API in which case a single type is quite natural. Other SDKs do the same thing we shouldn't differentiate between a Prefix and an actual Object.

Reply:
The isDir approach is needed in languages without sum types (i.e. the other SDKs, and also in the C standard library interface). It is a way to supply both directory and object values in a single type. In all those cases, every usage of ObjectInfo values needs to check isDir unless it is making specific assumptions. Properties like size, etag, object metadata are all meaningless and need to be set to meaningless zero values if it is of directory type. It is very weird to do it in Haskell (or also rust) which has sum types to neatly distinguish between these two types of values.

Tree walk implementations return a single type in low level interfaces to avoid checking if an entry is a directory (e.g. golang standard library) or because a streaming interface is desired and the language does not provide sum types. It is not necessary to do this for object storage, as list objects API response (the lowest level available interface) already clearly differentiates between objects and object prefixes. A higher level interface is simpler to use for the end user programmer, and we can easily do this in this SDK as we have sum types.

Reply:
I see the sum type behavior this allows for a different type style implementation. Now I see the point, perhaps you would also want to update the version for the project in this PR since this is a breaking change.